src/rt/lifetime.d

```diff
@@ -2591,33 +2591,75 @@ unittest
     }
 }
 
+// test class finalizers exception handling
+unittest
+{
+    bool test(E)()
+    {
+        import core.exception;
+        static class C1
+        {
+            // preallocate to not call new in destructor
+            __gshared Exception exc = new Exception("test onFinalizeError");
+            ~this()
+            {
+                throw exc;
+            }
+        }
+
+        bool caught = false;
+        C1 c = new C1;
+        try
+        {
+            GC.runFinalizers((cast(char*)c)[0..1]);
+        }
+        catch(FinalizeError err)
+        {
+            caught = true;
+        }
+        GC.free(cast(void*)c);
+        return caught;
+    }
+
+    assert( test!Exception);
+    import core.exception : InvalidMemoryOperationError;
+    assert(!test!InvalidMemoryOperationError);
```

Comment:
I guess you'll have to catch the memory error to successfully pass the unittest.